### PR TITLE
Implement the ability to upload a directory

### PR DESCRIPTION
**Pull request by arianna-h-h**

### CODE REVIEW

File: Main.js [hunk 1]
```diff
@@ -1,12 +1,30 @@
+const fs = require('fs');
 const { readFile } = require('./readFile.js');
 const { uploadFile } = require('./uploadFile.js');
+const { isFile } = require('./isFile.js');
 
-const gistName = process.argv[3];
-const file = process.argv[2];
+const contentAndNameList = process.argv;
```
Comment (snmgian) [inline Main.js]: always define variables in the narrowest scope as possible.

There 3 variables should be defined inside `Runner` function

File: folder/short2.txt [hunk 2]
```diff
@@ -0,0 +1,5 @@
+Hello!!!
+
+This is another short file.
+
+Files are fun.
```
Comment (snmgian) [inline folder/short2.txt:5]: :)

File: readFile.js [hunk 3]
```diff
@@ -4,26 +4,32 @@ const PROG_BAR = '=';
...
-        const prog = ((chunk.length / totalSize) * BAR_LENGTH);
-        process.stdout.write(`${PROG_BAR.repeat(Math.round(prog - HANG_PERCENT))}`);
-      });
-      readStream.on('close', () => {
-        resolve(finalString);
-      });
-    } catch (error) {
-      reject(error);
-    }
+  promiseList = filesToUpload.map((file) => {
```
Comment (snmgian) [inline readFile.js]: use `const` here

File: readFile.js [hunk 4]
```diff
@@ -4,26 +4,32 @@ const PROG_BAR = '=';
...
+    let finalContents = '';
+    return new Promise((resolve, reject) => {
+      try {
+        const readStream = fileSystem.createReadStream(file.content);
+        readStream.on('data', (chunk) => {
+          finalContents += chunk;
+          const prog = ((chunk.length / totalSize) * BAR_LENGTH);
```
Comment (snmgian) [inline readFile.js]: Move these calculations into a separate function

File: readFile.js [hunk 5]
```diff
@@ -4,26 +4,32 @@ const PROG_BAR = '=';
...
+        });
+        readStream.on('close', () => {
+          resolve({ [file.name]: { content: finalContents } });
+        });
+      } catch (error) {
+        reject(error);
+      }
+    });
   });
+  return Promise.all(promiseList).then(response => response);
```
Comment (snmgian) [inline readFile.js]: `Promise.all(promiseList).then(response => response)` is exactly the same as `Promise.all(promiseList)`

File: readFile.js [hunk 6]
```diff
@@ -4,26 +4,32 @@ const PROG_BAR = '=';
 const BAR_LENGTH = 30;
 const HANG_PERCENT = 0.25;
 
-
-function readFile(fileToContent) {
-  let finalString;
-  const totalSize = fileSystem.statSync(process.argv[2]).size;
+async function readFile(filesToUpload) {
+  let promiseList = [];
```
Comment (snmgian) [inline readFile.js]: Rename it as `promises`

**Review by snmgian — CHANGES_REQUESTED** [review]

File: Main.js (hunk 1 above)
Comment (arianna-h-h) [reply Main.js]: done

File: readFile.js (hunk 6 above)
Comment (arianna-h-h) [reply readFile.js]: done

File: readFile.js (hunk 3 above)
Comment (arianna-h-h) [reply readFile.js]: done

File: readFile.js (hunk 5 above)
Comment (arianna-h-h) [reply readFile.js]: done

File: readFile.js (hunk 4 above)
Comment (arianna-h-h) [reply readFile.js]: done

File: Main.js [hunk 7]
```diff
@@ -1,13 +1,30 @@
...
 
 async function Runner() {
-  const content = await readFile(file);
-  const response = await uploadFile(content, gistName);
+  const contentAndNameList = process.argv;
+  const objectList = [];
+  const serializedList = [];
+
+  for (let i = 2; i < contentAndNameList.length; i += 1) {
+    objectList.push({ name: contentAndNameList[i], content: null });
```
Comment (ashleymichal) [inline Main.js]: why have this null value?

File: Main.js [hunk 8]
```diff
@@ -1,13 +1,30 @@
...
+    if (isFile(item.name)) {
+      serializedList.push(item);
+    } else {
+      fs.readdirSync(item.name).forEach(fileFromDir =>
+        serializedList.push({
+          name: `${item.name}/${fileFromDir}`,
+          content: null,
+        }));
+    }
+  });
```
Comment (ashleymichal) [inline Main.js]: this whole block looks an awful lot like a use case for `flatMap()`. There is a lodash package that allows you to import _just_ [flatMap](https://lodash.com/docs/4.17.4#flatMap), or you can experiment with using .reduce() to write your own implementation. Dealer's choice.

I also think that the block is concerned with preparing to read files, and as such might be a good fit for a helper function in that file.

File: gistUpload.test.js [hunk 9]
```diff
@@ -15,14 +15,14 @@ describe('When given no arguments or incorrect number of arguments', () => {
   it('Returns a new error message: you must provide a file and name.', () => {
     uploadFile()
       .catch((error) => {
-        expect(error.message).toEqual('You must provide a file and a name.');
+        expect(error.message).toEqual('You must provide a files.');
```
Comment (ashleymichal) [inline gistUpload.test.js]: `You must provide a files`

File: isFile.js [hunk 10]
```diff
@@ -0,0 +1,13 @@
...
+
+/** The isFile function determines if a file path is a file.
+  * @param {string} file - file or directory path.
+  * @return {boolean} - returns true if file is a file.
+  */
+
+function isFile(file) {
+  const stats = fs.statSync(file);
+  return stats.isFile();
+}
```
Comment (ashleymichal) [inline isFile.js]: this function is also a nice little helper function for read file.

**Review by ashleymichal — CHANGES_REQUESTED** [review]

File: Main.js (hunk 7 above)
Comment (arianna-h-h) [reply Main.js]: done

File: gistUpload.test.js (hunk 9 above)
Comment (arianna-h-h) [reply gistUpload.test.js]: done

File: Main.js (hunk 8 above)
Comment (arianna-h-h) [reply Main.js]: done

File: isFile.js (hunk 10 above)
Comment (arianna-h-h) [reply isFile.js]: done

File: Main.js [hunk 11]
```diff
@@ -1,28 +1,11 @@
-const fs = require('fs');
 const readFile = require('./readFile.js');
 const uploadFile = require('./uploadFile.js');
 const clearScreen = require('./clearScreen.js');
 const isFile = require('./isFile.js');
 
 async function Runner() {
   const contentAndNameList = process.argv;
```
Comment (ashleymichal) [inline Main.js]: slice this 😁

File: isFile.js [hunk 12]
```diff
@@ -1,13 +1,31 @@
...
   * @param {string} file - file or directory path.
   * @return {boolean} - returns true if file is a file.
   */
 
-function isFile(file) {
-  const stats = fs.statSync(file);
-  return stats.isFile();
+function isFile(objectList) {
+  objectList.shift();
+  objectList.shift();
```
Comment (ashleymichal) [inline isFile.js]: these will become extraneous when you slice your args.

File: isFile.js [hunk 13]
```diff
@@ -1,13 +1,31 @@
...
+  if (stats.isFile()) {
+    serializedList.push({ name: arg });
+  } else {
+    fs.readdirSync(arg).forEach(fileFromDir =>
+      serializedList.push({
+        name: `${arg}/${fileFromDir}`,
+      }));
+  }
+  return serializedList;
+}
```
Comment (ashleymichal) [inline isFile.js]: I think maybe you are just re-implementing flatmap here...

File: Main.js [hunk 14]
```diff
@@ -1,28 +1,11 @@
...
-      serializedList.push(item);
-    } else {
-      fs.readdirSync(item.name).forEach(fileFromDir =>
-        serializedList.push({
-          name: `${item.name}/${fileFromDir}`,
-          content: null,
-        }));
-    }
-  });
+  const serializedList = await isFile(contentAndNameList);
```
Comment (ashleymichal) [inline Main.js]: I don't think this is an async function.

**Review by ashleymichal — CHANGES_REQUESTED** [review]

File: Main.js (hunk 11 above)
Comment (arianna-h-h) [reply Main.js]: done

File: Main.js (hunk 14 above)
Comment (arianna-h-h) [reply Main.js]: done

File: isFile.js (hunk 12 above)
Comment (arianna-h-h) [reply isFile.js]: done

File: isFile.js (hunk 13 above)
Comment (arianna-h-h) [reply isFile.js]: done

File: gistUpload.test.js [hunk 15]
```diff
@@ -15,14 +15,14 @@ describe('When given no arguments or incorrect number of arguments', () => {
   it('Returns a new error message: you must provide a file and name.', () => {
     uploadFile()
       .catch((error) => {
-        expect(error.message).toEqual('You must provide a file and a name.');
+        expect(error.message).toEqual('You must provide files.');
```
Comment (snmgian) [inline gistUpload.test.js]: Simplify the string being passed to `it`

File: gistUpload.test.js [hunk 16]
```diff
@@ -15,14 +15,14 @@ describe('When given no arguments or incorrect number of arguments', () => {
...
       });
   });
 
   it('Returns a new error message: you must provide a file and name.', () => {
     uploadFile('hello')
       .catch((error) => {
-        expect(error.message).toEqual('You must provide a file and a name.');
+        expect(error.message).toEqual('You must provide files.');
```
Comment (snmgian) [inline gistUpload.test.js]: Idem

File: gistUpload.test.js [hunk 17]
```diff
@@ -31,14 +31,14 @@ describe('When given non-string arguments', () => {
   it('Returns the error: Your file must contain a string.', () => {
     uploadFile(1)
       .catch((error) => {
-        expect(error.message).toEqual('Your file must contain a string.');
+        expect(error.message).toEqual('You must provide files.');
```
Comment (snmgian) [inline gistUpload.test.js]: Idem

File: gistUpload.test.js [hunk 18]
```diff
@@ -31,14 +31,14 @@ describe('When given non-string arguments', () => {
...
       });
   });
 
   it('Returns the error: Your gist name must be a string.', () => {
     uploadFile('hello', 2)
       .catch((error) => {
-        expect(error.message).toEqual('Your gist name must be a string.');
+        expect(error.message).toEqual('You must provide files.');
```
Comment (snmgian) [inline gistUpload.test.js]: Idem

File: package.json [hunk 19]
```diff
@@ -17,6 +17,7 @@
   "dependencies": {
     "axios": "^0.16.2",
     "dotenv": "^4.0.0",
+    "lodash.flatmap": "^4.5.0",
```
Comment (snmgian) [inline package.json:20]: Why not adding lodash instead of lodash.flatmap ?

File: readFile.js [hunk 20]
```diff
@@ -1,38 +1,52 @@
...
-      reject(error);
-    }
+  const promises = filesToUpload.map((file) => {
+    let finalContents = '';
+    return new Promise((resolve, reject) => {
+      try {
+        const readStream = fileSystem.createReadStream(file.name);
+        readStream.on('data', (chunk) => {
+          finalContents += chunk;
+          const percent = (Math.round((finalContents.length / totalSize) * 100));
```
Comment (snmgian) [inline readFile.js]: Move these lines into a separate function

**Review by snmgian — CHANGES_REQUESTED** [review]
Also, remove yarn-error.log and add it into gitignore

File: package.json (hunk 19 above)
Comment (arianna-h-h) [reply package.json:20]: Because I'm only using flatmap

**Review by snmgian — APPROVED** [review]

**Pull request merged**